### PR TITLE
Update to fcl 0.6.1 and ci build for linux/win/mac using github actions

### PR DESCRIPTION
**Pull request by CyrilWaechter**
This pull request kind of combine #21 and #18:
* Update to `0.6.1` with #21 but fixed to be cross-platform
* Add windows build like #18 but:
  * based on `fcl 0.6.1`
  * same `setup.py` for all platform
  * add `setup.cfg` for static metadata and `pyproject.toml` to fine tune ci options
  * platform specific build script for dependencies
  * using github actions as ci provider to allow build simultaneously on linux/mac/windows.

### CODE REVIEW

**Comment by mikedh** [comment]
This is awesome! I checked with @mmatl a while back and he was cool with me merging one of the PR's updating to the latest version of FCL and the cross-platform stuff is also awesome. This looks like it did the work to combine those PR's and check via multiplatform builds. I'm happy to review/test/merge once builds/tests/conflicts are resolved, unless someone has strong objections.

**Comment by mmatl** [comment]
Thanks for the big PR! @mikedh @CyrilWaechter just let me know when this is fully ready and I can stage a deploy

**Comment by CyrilWaechter** [comment]
@mikedh @mmatl  I did not expect an answer. Happy to see that the repo is more alive than I thought :smiley: .

* I have rebased my branch on top of master to allow cleaner merge
* I have udated readme, setup.cfg, dockerfile and travis (in case you keep travis)
* I have split github actions workflows to upload wheels only on release

Let me know if you want me to fix/change something else.

**Review by mmatl — APPROVED** [review]

**Comment by mmatl** [comment]
@CyrilWaechter Looks good to me! I'll go ahead and merge it now.

**Comment by mikedh** [comment]
It's not dead, it's just resting it's eyes haha. Getting those PR's across the line is a ton of work, thanks for taking it on!

I installed the wheels generated via this PR built for 3.8/manylinux, and with one minor issue my upstream tests worked. I did have to change data I was passing from `float32` to `float64` to get upstream tests to pass:

```
Traceback (most recent call last):
  File "/home/mikedh/trimesh/trimesh/collision.py", line 204, in add_object
    t = fcl.Transform(transform[:3, :3], transform[:3, 3])
ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
Exception ignored in: 'fcl.fcl.numpy_to_vec3d'
```
If `fcl.numpy_to_vec32` requires a float64/double, should it always convert in that function via a `.astype`?

Only other thought is I think we should probably drop Travis completely and just use Github actions.

**Comment by mmatl** [comment]
@mjd3 Could you merge this PR and do a deploy? I no longer have access rights to this repo, it appears.

**Comment by CyrilWaechter** [comment]
> I installed the wheels generated via this PR built for 3.8/manylinux, and with one minor issue my upstream tests worked. I did have to change data I was passing from `float32` to `float64` to get upstream tests to pass:
> 
> ```
> Traceback (most recent call last):
>   File "/home/mikedh/trimesh/trimesh/collision.py", line 204, in add_object
>     t = fcl.Transform(transform[:3, :3], transform[:3, 3])
> ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
> ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
> Exception ignored in: 'fcl.fcl.numpy_to_vec3d'
> ```
> 
> If `fcl.numpy_to_vec32` requires a float64/double, should it always convert in that function via a `.astype`?

Sorry. I have no idea. @DmitryNeverov did the core work. I just fixed bvh path and work on the packaging side. On my side I just checked that all test were passing and that I was able to use clash detection tool from [BlenderBIM Add-on](https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.6.0/src/blenderbim) which currently use `python-fcl`.

File: src/fcl/fcl.pyx [hunk 1]
```diff
@@ -686,75 +722,76 @@ cdef class DistanceFunction:
...
+    return defs.Quaterniond(<double?> a[0], <double?> a[1], <double?> a[2], <double?> a[3])
 
-cdef vec3f_to_numpy(defs.Vec3f vec):
+cdef vec3d_to_numpy(defs.Vector3d vec):
     return numpy.array([vec[0], vec[1], vec[2]])
 
-cdef defs.Vec3f numpy_to_vec3f(a):
-    return defs.Vec3f(<double?> a[0], <double?> a[1], <double?> a[2])
+cdef defs.Vector3d numpy_to_vec3d(np.ndarray[double, ndim=1] a):
+    return defs.Vector3d(&a[0])
```
Comment (DmitryNeverov) [inline src/fcl/fcl.pyx:754]:
```suggestion
cdef defs.Vector3d numpy_to_vec3d(arr):
    cdef double[:] memview = arr.astype(numpy.float64)
    return defs.Vector3d(&memview[0])
```

File: src/fcl/fcl.pyx [hunk 2]
```diff
@@ -686,75 +722,76 @@ cdef class DistanceFunction:
...
 
-cdef defs.Matrix3f numpy_to_mat3f(a):
-    return defs.Matrix3f(<double?> a[0][0], <double?> a[0][1], <double?> a[0][2],
-                         <double?> a[1][0], <double?> a[1][1], <double?> a[1][2],
-                         <double?> a[2][0], <double?> a[2][1], <double?> a[2][2])
+cdef defs.Matrix3d numpy_to_mat3d(np.ndarray[double, ndim=2] a):
+    # NOTE Eigen defaults to column-major storage,
+    # which corresponds to non-default Fortran mode of ordering in numpy
+    cdef np.ndarray[double, ndim=2, mode='fortran'] f = np.ndarray.copy(a, order='F')
+    return defs.Matrix3d(&f[0, 0])
```
Comment (DmitryNeverov) [inline src/fcl/fcl.pyx:765]:
```suggestion
cdef defs.Matrix3d numpy_to_mat3d(arr):
    # NOTE Eigen defaults to column-major storage,
    # which corresponds to non-default Fortran mode of ordering in numpy
    cdef double[:, :] memview = arr.astype(numpy.float64, order='F')
    return defs.Matrix3d(&memview[0, 0])
```

**Comment by DmitryNeverov** [comment]
@CyrilWaechter 
Glad to see this eventually got some traction, thanks for pushing it forward.

I've left suggestions above that should fix
```
ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
```
that @mikedh was facing.

**Comment by mikedh** [comment]
Awesome thanks @DmitryNeverov @Pebaz @CyrilWaechter for the fixes and all the work on the upgrade! I just checked the regenerated wheels and they fixed the only minor issue I was seeing. It looks like I still have merge privileges on this repo and I'm happy to merge this (I think it's ready to go), although I think the CI gnomes won't release to PyPi correctly until we do the following:

- @mmatl could you generate a PyPi repo access token for `python-fcl`?
- Could someone with admin on this repo (@mjd3?) add that token as `PYPI_TOKEN` (or something) secret to Github Actions? It looks like I can merge stuff here but not add secrets. 
- I'll delete .travis.ci and add the PyPi release step in a followup PR. 

That sound OK?

**Comment by mmatl** [comment]
@mikedh Sorry for the delay! Token has been generated, and @mjd3 will take care of getting it added to the repo's secrets. Feel free to perform the merge and add the github actions PR as soon as you have time.

**Comment by mjd3** [comment]
@mikedh token is added as `PYPI_API_TOKEN` to the repo's secrets. Let me know if you need any assistance with the GH Actions PR (although you probably have more experience with that than me)!

**Comment by mikedh** [comment]
Awesome thanks @mjd3 @mmatl! I'll block off an afternoon and get this released, I'll let you know if I run in to trouble haha.

**Comment by szobov** [comment]
Hi there,
Thanks a lot for this update,
Any chance it will be released soon?
Cheers

**Pull request merged**